sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -177,14 +178,44 @@ trait SpillableHandle extends StoreHandle {
    */
   def spill(): Long
 
+  private val spilling = new AtomicBoolean(false)
+
+  /**
+   * Method used to atomically check and set the spilling state, so that anyone who wants to
+   * actually perform a spill can ensure they are the only one spilling, without having to block
+   * on the actual spill operation (IO). Only someone who has set spilling to true to perform their
+   * spill may set it back to false when they are done. (Visible for tests)
+   *
+   * This is a separate check from spillable, which actually checks the state of the buffer handle
+   *
+   * @param s whether the caller is trying to spill or not (ie finished)
+   * @return whether the caller is allowed to spill (or true if s is false)
+   */
+  def setSpilling(s: Boolean): Boolean = {
+    if (!s) {
+      if (!spilling.getAndSet(false)) {
+        throw new IllegalStateException("tried to setSpilling to false while not spilling!")
```

Comment:
Why throw in this scenario?

Reply:
Someone should only be able to mark that we're no longer spilling in the case that we are currently spilling/they have the spill lock. If that's not the case, there's likely a bug, and it could result in us thinking we're not spilling when we are, leading to multiple threads spilling the same buffer.